File: ➡️📂Core kernel Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/nodejs/test/e2e/multi_provider_registry.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import type {
    CustomAgentConfig,
    NamedProviderConfig,
    ProviderModelConfig,
} from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";
import { retry } from "./harness/sdkTestHelper.js";
import type { ParsedHttpExchange } from "../../../test/harness/replayingCapiProxy";

/**
 * End-to-end coverage for the experimental multi-provider BYOK registry
 * (`providers` / `models` on the session config). Validates that several named
 * providers, several models per provider, and custom agents bound to those
 * provider-qualified models can coexist in one session, be launched, and route
 * inference to the configured provider with the configured wire model and
 * headers.
 */
describe("Multi-provider BYOK registry", async () => {
    const { copilotClient: client, openAiEndpoint } = await createSdkTestContext();

    async function waitForExchanges(minimumCount = 1): Promise<ParsedHttpExchange[]> {
        await retry(
            `capture ${minimumCount} chat completion request(s)`,
            async () => {
                const exchanges = await openAiEndpoint.getExchanges();
                expect(exchanges.length).toBeGreaterThanOrEqual(minimumCount);
            },
            1_200
        );
        return openAiEndpoint.getExchanges();
    }

    function getHeader(exchange: ParsedHttpExchange, name: string): string | undefined {
        const headers = exchange.requestHeaders ?? {};
        const key = Object.keys(headers).find((k) => k.toLowerCase() === name.toLowerCase());
        if (key === undefined) {
            return undefined;
        }
        const value = headers[key];
        return Array.isArray(value) ? value[0] : value;
    }

    // A heterogeneous registry: two providers of different types, with multiple
    // models each. Provider-qualified selection ids are alpha/sonnet,
    // alpha/haiku, beta/opus, beta/haiku.
    const registryProviders: NamedProviderConfig[] = [
        {
            name: "alpha",
            type: "openai",
            wireApi: "completions",
            baseUrl: "https://alpha.example.test/v1",
            apiKey: "alpha-secret",
            headers: { "X-Provider": "alpha" },
        },
        {
            name: "beta",
            type: "anthropic",
            baseUrl: "https://beta.example.test",
            bearerToken: "beta-bearer",
            headers: { "X-Provider": "beta" },
        },
    ];
    const registryModels: ProviderModelConfig[] = [
        { id: "sonnet", provider: "alpha", wireModel: "byok-gpt-4o", maxPromptTokens: 111111 },
        { id: "haiku", provider: "alpha", wireModel: "byok-gpt-4o-mini" },
        { id: "opus", provider: "beta", wireModel: "byok-claude-3-opus" },
        { id: "haiku", provider: "beta", wireModel: "byok-claude-3-haiku" },
    ];
    const registryAgents: CustomAgentConfig[] = [
        {
            name: "orchestrator",
            displayName: "Orchestrator",
            description: "Top-level planner.",
            prompt: "Plan and delegate.",
            model: "alpha/sonnet",
        },
        {
            name: "researcher",
            displayName: "Researcher",
            description: "Deep research subagent.",
            prompt: "Research thoroughly.",
            model: "beta/opus",
        },
        {
            name: "fast-helper",
            displayName: "Fast Helper",
            description: "Quick subagent.",
            prompt: "Answer quickly.",
            model: "alpha/haiku",
        },
        {
            name: "summarizer",
            displayName: "Summarizer",
            description: "Summarizing subagent.",
            prompt: "Summarize.",
            model: "beta/haiku",
        },
    ];

    it("should register multiple providers with custom agents bound to their models", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            providers: registryProviders,
            models: registryModels,
            customAgents: registryAgents,
        });

        try {
            const { agents } = await session.rpc.agent.list();

            // All four custom agents coexist in a single session.
            expect(agents.length).toBe(4);

            // Each agent is bound to its configured provider-qualified BYOK model.
            const byName = new Map(agents.map((a) => [a.name, a]));
            expect(byName.get("orchestrator")?.model).toBe("alpha/sonnet");
            expect(byName.get("researcher")?.model).toBe("beta/opus");
            expect(byName.get("fast-helper")?.model).toBe("alpha/haiku");
            expect(byName.get("summarizer")?.model).toBe("beta/haiku");

            // Models from BOTH providers are represented, proving the two
            // providers and their models coexist within the same session.
            const boundModels = agents.map((a) => a.model ?? "");
            expect(boundModels.some((m) => m.startsWith("alpha/"))).toBe(true);
            expect(boundModels.some((m) => m.startsWith("beta/"))).toBe(true);
        } finally {
            await session.disconnect();
        }
    });

    async function assertRouting(
        selectionId: string,
        expectedWireModel: string,
        expectedProviderHeader: string
    ): Promise<void> {
        // Two OpenAI-compatible providers, both pointed at the replay proxy so
        // their /chat/completions traffic is captured. They are distinguished on
        // the wire by their per-provider X-Provider header. "alpha" carries two
        // models (multiple models per provider); "delta" carries one.
        const providers: NamedProviderConfig[] = [
            {
                name: "alpha",
                type: "openai",
                wireApi: "completions",
                baseUrl: openAiEndpoint.url,
                apiKey: "alpha-secret",
                headers: { "X-Provider": "alpha" },
            },
            {
                name: "delta",
                type: "openai",
                wireApi: "completions",
                baseUrl: openAiEndpoint.url,
                apiKey: "delta-secret",
                headers: { "X-Provider": "delta" },
            },
        ];
        const models: ProviderModelConfig[] = [
            { id: "sonnet", provider: "alpha", wireModel: "byok-gpt-4o" },
            { id: "haiku", provider: "alpha", wireModel: "byok-gpt-4o-mini" },
            { id: "turbo", provider: "delta", wireModel: "byok-gpt-4-turbo" },
        ];

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            model: selectionId,
            providers,
            models,
        });

        try {
            await session.sendAndWait({ prompt: "What is 5+5?" });
            const exchanges = await waitForExchanges();
            expect(exchanges.length).toBe(1);
            const exchange = exchanges[0];

            // The wire model sent to the provider is the selected model's
            // wireModel, not its provider-qualified selection id.
            expect(exchange.request.model).toBe(expectedWireModel);

            // The request carried the owning provider's custom header, proving
            // the turn was dispatched against the correct provider connection.
            expect(getHeader(exchange, "X-Provider")).toBe(expectedProviderHeader);

            // The provider's API key was applied as an Authorization header.
            expect(getHeader(exchange, "Authorization")).toBeTruthy();
        } finally {
            try {
                await session.disconnect();
            } catch {
                // disconnect may fail since the BYOK provider URL is fake
            }
        }
    }

    it("should route alpha sonnet turn to its provider and wire model", async () => {
        await assertRouting("alpha/sonnet", "byok-gpt-4o", "alpha");
    });

    it("should route alpha haiku turn to its provider and wire model", async () => {
        await assertRouting("alpha/haiku", "byok-gpt-4o-mini", "alpha");
    });

    it("should route delta turbo turn to its provider and wire model", async () => {
        await assertRouting("delta/turbo", "byok-gpt-4-turbo", "delta");
    });
});

File: ➡️📂Core kernel Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/go/internal/e2e/multi_provider_registry_e2e_test.go
package e2e

import (
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

// TestMultiProviderRegistryE2E exercises the experimental multi-provider BYOK
// registry (Providers / Models on the session config). It validates that
// several named providers, several models per provider, and custom agents
// bound to those provider-qualified models can coexist in one session, be
// launched, and route inference to the configured provider with the configured
// wire model and headers.
func TestMultiProviderRegistryE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	t.Run("should register multiple providers with custom agents bound to their models", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		// A heterogeneous registry: two providers of different types, with
		// multiple models each. Provider-qualified selection ids are
		// alpha/sonnet, alpha/haiku, beta/opus, beta/haiku.
		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Providers: []copilot.NamedProviderConfig{
				{
					Name:    "alpha",
					Type:    "openai",
					WireAPI: "completions",
					BaseURL: "https://alpha.example.test/v1",
					APIKey:  "alpha-secret",
					Headers: map[string]string{"X-Provider": "alpha"},
				},
				{
					Name:        "beta",
					Type:        "anthropic",
					BaseURL:     "https://beta.example.test",
					BearerToken: "beta-bearer",
					Headers:     map[string]string{"X-Provider": "beta"},
				},
			},
			Models: []copilot.ProviderModelConfig{
				{ID: "sonnet", Provider: "alpha", WireModel: "byok-gpt-4o", MaxPromptTokens: 111111},
				{ID: "haiku", Provider: "alpha", WireModel: "byok-gpt-4o-mini"},
				{ID: "opus", Provider: "beta", WireModel: "byok-claude-3-opus"},
				{ID: "haiku", Provider: "beta", WireModel: "byok-claude-3-haiku"},
			},
			CustomAgents: []copilot.CustomAgentConfig{
				{Name: "orchestrator", DisplayName: "Orchestrator", Description: "Top-level planner.", Prompt: "Plan and delegate.", Model: "alpha/sonnet"},
				{Name: "researcher", DisplayName: "Researcher", Description: "Deep research subagent.", Prompt: "Research thoroughly.", Model: "beta/opus"},
				{Name: "fast-helper", DisplayName: "Fast Helper", Description: "Quick subagent.", Prompt: "Answer quickly.", Model: "alpha/haiku"},
				{Name: "summarizer", DisplayName: "Summarizer", Description: "Summarizing subagent.", Prompt: "Summarize.", Model: "beta/haiku"},
			},
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}

		result, err := session.RPC.Agent.List(t.Context())
		if err != nil {
			t.Fatalf("Agent.List failed: %v", err)
		}

		// All four custom agents coexist in a single session.
		if len(result.Agents) != 4 {
			t.Fatalf("Expected 4 agents, got %d", len(result.Agents))
		}

		// Each agent is bound to its configured provider-qualified BYOK model.
		boundModels := map[string]string{}
		for _, agent := range result.Agents {
			model := ""
			if agent.Model != nil {
				model = *agent.Model
			}
			boundModels[agent.Name] = model
		}
		expected := map[string]string{
			"orchestrator": "alpha/sonnet",
			"researcher":   "beta/opus",
			"fast-helper":  "alpha/haiku",
			"summarizer":   "beta/haiku",
		}
		for name, want := range expected {
			if got := boundModels[name]; got != want {
				t.Errorf("Expected agent %q bound to model %q, got %q", name, want, got)
			}
		}

		// Models from BOTH providers are represented, proving the two providers
		// and their models coexist within the same session.
		var hasAlpha, hasBeta bool
		for _, model := range boundModels {
			if strings.HasPrefix(model, "alpha/") {
				hasAlpha = true
			}
			if strings.HasPrefix(model, "beta/") {
				hasBeta = true
			}
		}
		if !hasAlpha || !hasBeta {
			t.Errorf("Expected both providers represented; hasAlpha=%v hasBeta=%v", hasAlpha, hasBeta)
		}
	})

	assertRouting := func(t *testing.T, selectionID, expectedWireModel, expectedProviderHeader string) {
		ctx.ConfigureForTest(t)

		// Two OpenAI-compatible providers, both pointed at the replay proxy so
		// their /chat/completions traffic is captured. They are distinguished
		// on the wire by their per-provider X-Provider header. "alpha" carries
		// two models (multiple models per provider); "delta" carries one.
		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Model:               selectionID,
			Providers: []copilot.NamedProviderConfig{
				{
					Name:    "alpha",
					Type:    "openai",
					WireAPI: "completions",
					BaseURL: ctx.ProxyURL,
					APIKey:  "alpha-secret",
					Headers: map[string]string{"X-Provider": "alpha"},
				},
				{
					Name:    "delta",
					Type:    "openai",
					WireAPI: "completions",
					BaseURL: ctx.ProxyURL,
					APIKey:  "delta-secret",
					Headers: map[string]string{"X-Provider": "delta"},
				},
			},
			Models: []copilot.ProviderModelConfig{
				{ID: "sonnet", Provider: "alpha", WireModel: "byok-gpt-4o"},
				{ID: "haiku", Provider: "alpha", WireModel: "byok-gpt-4o-mini"},
				{ID: "turbo", Provider: "delta", WireModel: "byok-gpt-4-turbo"},
			},
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}

		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 5+5?"}); err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}

		exchanges, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("GetExchanges failed: %v", err)
		}
		if len(exchanges) != 1 {
			t.Fatalf("Expected exactly 1 exchange, got %d", len(exchanges))
		}
		exchange := exchanges[0]

		// The wire model sent to the provider is the selected model's WireModel,
		// not its provider-qualified selection id.
		if exchange.Request.Model != expectedWireModel {
			t.Errorf("Expected request model %q, got %q", expectedWireModel, exchange.Request.Model)
		}

		// The request carried the owning provider's custom header, proving the
		// turn was dispatched against the correct provider connection.
		if !exchangeHasHeader(exchange, "X-Provider", expectedProviderHeader) {
			t.Errorf("Expected X-Provider header %q to be present", expectedProviderHeader)
		}

		// The provider's API key was applied as an Authorization header.
		if !exchangeHasHeader(exchange, "Authorization", "Bearer") {
			t.Error("Expected an Authorization header on the dispatched request")
		}
	}

	t.Run("should route alpha sonnet turn to its provider and wire model", func(t *testing.T) {
		assertRouting(t, "alpha/sonnet", "byok-gpt-4o", "alpha")
	})

	t.Run("should route alpha haiku turn to its provider and wire model", func(t *testing.T) {
		assertRouting(t, "alpha/haiku", "byok-gpt-4o-mini", "alpha")
	})

	t.Run("should route delta turbo turn to its provider and wire model", func(t *testing.T) {
		assertRouting(t, "delta/turbo", "byok-gpt-4-turbo", "delta")
	})
}
